internal/transport/transport_test.go

```diff
@@ -3204,3 +3206,141 @@ func (s) TestClientTransport_Handle1xxHeaders(t *testing.T) {
+		channelz.TurnOn()
+	}
+
+	for _, test := range []struct {
+		name                string
+		eosReceived         bool
+		removeFromActive    bool // Remove stream from activeStreams before calling deleteStream
+		wantStreamSucceeded int64
+		wantStreamFailed    int64
+	}{
+		{
+			name:                "StreamsSucceeded",
+			eosReceived:         true,
+			removeFromActive:    false,
+			wantStreamSucceeded: 1,
+			wantStreamFailed:    0,
+		},
+		{
+			name:                "StreamsFailed",
+			eosReceived:         false,
+			removeFromActive:    false,
+			wantStreamSucceeded: 0,
+			wantStreamFailed:    1,
+		},
+		{
+			name:                "NonActiveStream",
+			eosReceived:         true,
+			removeFromActive:    true, // Stream already removed from activeStreams
+			wantStreamSucceeded: 0,
+			wantStreamFailed:    0,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			// Setup server configuration with channelz support
+			serverConfig := &ServerConfig{
+				ChannelzParent: channelz.RegisterServer("test server"),
+			}
+			defer channelz.RemoveEntry(serverConfig.ChannelzParent.ID)
+
+			// Create server and client with normal handler (not notifyCall)
+			server, client, cancel := setUpWithOptions(t, 0, serverConfig, normal, ConnectOptions{})
+			defer cancel()
+			defer server.stop()
+			defer client.Close(fmt.Errorf("test cleanup"))
+
+			// Wait for connection to be established
+			waitWhileTrue(t, func() (bool, error) {
+				server.mu.Lock()
+				defer server.mu.Unlock()
+				if len(server.conns) == 0 {
+					return true, fmt.Errorf("timed-out while waiting for connection")
+				}
+				return false, nil
+			})
+
+			// Get the server transport
+			server.mu.Lock()
+			var serverTransport *http2Server
+			for st := range server.conns {
+				serverTransport = st.(*http2Server)
+				break
+			}
+			server.mu.Unlock()
+
+			if serverTransport == nil {
+				t.Fatal("Server transport not found")
+			}
+
+			// Always create a real stream through the client
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			clientStream, err := client.NewStream(ctx, &CallHdr{})
+			if err != nil {
+				t.Fatalf("Failed to create stream: %v", err)
+			}
+
+			// Wait for the stream to be created on the server side
+			var serverStream *ServerStream
+			waitWhileTrue(t, func() (bool, error) {
+				serverTransport.mu.Lock()
+				defer serverTransport.mu.Unlock()
+				for _, v := range serverTransport.activeStreams {
+					if v.id == clientStream.id {
+						serverStream = v
+						return false, nil
+					}
+				}
+				return true, nil
+			})
+
+			if serverStream == nil {
+				t.Fatalf("Server stream not found for client stream ID %d", clientStream.id)
+			}
+
+			// For the NonActiveStream test, remove the stream from activeStreams
+			if test.removeFromActive {
+				serverTransport.mu.Lock()
+				delete(serverTransport.activeStreams, serverStream.id)
+				serverTransport.mu.Unlock()
+			}
```

Comment:
I think we should not access the `activeStreams` map directly. We should rely on the first call to `serverTransport.deleteStream` to delete the stream. After the first call to `serverTransport.deleteStream`, we should assert that the metric counts are as expected.

After the first call to `serverTransport.deleteStream`, we can assume that the stream is deleted from the `activeStreams` map. We can call `serverTransport.deleteStream` a couple of times more and verify that the metric values haven't changed.

This would allow us to get rid of the `removeFromActive` test param since that case is already covered by the assertions made in the previous paragraph.

Reply:
[252532e](https://github.com/grpc/grpc-go/pull/8573/commits/252532ea6eab5e151c95a2bc7b8aa7270c050fd0)